<a href="https://colab.research.google.com/github/coolindye/cse380-notebooks/blob/master/Copy_of_about_not_so_basic_probability_theory.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# About Not So Basic Probability Theory
## Divide Pair Conquer
### Due: Monday, 8 February 2021

## Euler's Famous Sum

This famous problem that no mathematician until Euler knew how to solve involves summing the reciprocals of squares:
$$\sum^{\infty}_{n=1} \frac{1}{n^2} = \left(\frac{1}{1} + \frac{1}{4} + \frac{1}{9} + \frac{1}{16} + \frac{1}{25} + \cdots\right)$$


In [1]:
def compute_Eulers_famous_sum(max):
  return sum((1.0 / (n * n) for n in range(1, max + 1)))

In [2]:
print(list(map(compute_Eulers_famous_sum, range(1, 100))))

[1.0, 1.25, 1.3611111111111112, 1.4236111111111112, 1.4636111111111112, 1.4913888888888889, 1.511797052154195, 1.527422052154195, 1.5397677311665408, 1.5497677311665408, 1.558032193976458, 1.5649766384209025, 1.5708937981842162, 1.5759958390005426, 1.580440283444987, 1.584346533444987, 1.587806741057444, 1.5908931608105303, 1.5936632439130234, 1.5961632439130233, 1.5984308176091684, 1.6004969333116477, 1.6023872924798896, 1.6041234035910008, 1.6057234035910009, 1.6072026935318293, 1.6085744356443121, 1.6098499458483937, 1.6110390064904865, 1.6121501176015975, 1.6131907003279242, 1.6141672628279242, 1.6150855364734706, 1.6159505883765848, 1.616766914907197, 1.6175385198454686, 1.6182689800353882, 1.6189615008110114, 1.619618963006935, 1.6202439630069352, 1.6208388470045556, 1.621405740428592, 1.6219465733112313, 1.6224631022368512, 1.622956929397345, 1.6234295191894055, 1.623882212715888, 1.6243162404936657, 1.6247327336215291, 1.625132733621529, 1.6255172011340242, 1.6258870236192313, 

In [3]:
print(list(map(compute_Eulers_famous_sum, range(99997, 100000))))

[1.6449240665982363, 1.6449240666982403, 1.6449240667982423]


## Here's what Euler proved

The sum converges to exactly $\frac{\pi^2}{6}$.

In [ ]:
from math import pi
(pi ** 2) / 6

## How did he do it?

He used a **polynomial principle** --- namely, factorization.

For example:
$$P(x) = \left[1 - x + \frac{11x^2}{36} - \frac{x^3}{36} \right]$$

factors as

$$\left[\left(1 - \frac{x}{2}\right)\right] \left[\left(1 - \frac{x}{3}\right)\right] \left[\left(1 - \frac{x}{6}\right)\right]$$

and hence, this **finite** polynomial has roots at 2, 3, and 6.

## Using Trigonmetry and Calculus

Euler claimed the same principle applies to **infinite** polynomials, e.g.:
$$sin(x) = x - \frac{x^3}{3!} + \frac{x^5}{5!} - \frac{x^7}{7!} \pm \cdots$$

## And if x does not equal zero:

$$\frac{sin(x)}{x} = 1 - \frac{x^2}{3!} + \frac{x^4}{5!} - \frac{x^6}{7!} \pm \cdots$$

## which equals

$$\left[\left(1 - \frac{x}{\pi}\right)\left(1 - \frac{x}{-\pi}\right)\right] \left[\left(1 - \frac{x}{2\pi}\right)\left(1 - \frac{x}{-2\pi}\right)\right] \left[\left(1 - \frac{x}{3\pi}\right)\left(1 - \frac{x}{-3\pi}\right)\right] \cdots$$

which

## in turn equals

$$\left[\left(1 - \frac{x}{\pi}\right)\left(1 + \frac{x}{\pi}\right)\right] \left[\left(1 - \frac{x}{2\pi}\right)\left(1 + \frac{x}{2\pi}\right)\right] \left[\left(1 - \frac{x}{3\pi}\right)\left(1 + \frac{x}{3\pi}\right)\right] \cdots$$



## which equals

$$\left[\left(1 - \frac{x^2}{\pi^2}\right)\right] \left[\left(1 - \frac{x^2}{4\pi^2}\right)\right] \left[\left(1 - \frac{x^2}{9\pi^2}\right)\right]\cdots$$

## which equals

$$\left[1 - x^2 \left(\frac{1}{\pi^2} + \frac{1}{4\pi^2} + \frac{1}{9\pi^2} + \frac{1}{16\pi^2} + \cdots \right)\right] + \mbox{stuff}$$

## which equals

$$\left[1 - \frac{x^2}{\pi^2} \left(\frac{1}{1} + \frac{1}{4} + \frac{1}{9} + \frac{1}{16} + \cdots \right)\right] + \mbox{stuff}$$

## and therefore

$$1 - \frac{x^2}{6} = 1 - \frac{x^2}{\pi^2} \left(\frac{1}{1} + \frac{1}{4} + \frac{1}{9} + \frac{1}{16} + \cdots \right)$$

## and hence

$$\left(\frac{1}{1} + \frac{1}{4} + \frac{1}{9} + \frac{1}{16} + \cdots \right) = \frac{\pi^2}{6}.$$

**Do you see it?!**

## More Surprises in Store

### Coprime Probability
What is the probability that two randomly chosen positive integers will be coprime?

Recall the definition of *coprime*, or *relatively prime*. Two numbers are coprime if their greatest common divisor is 1, as is the case with any two *prime* numbers.

A prime, $p$, is coprime with any other positive integer that is not a multiple of $p$.


### But does at least one of the two numbers have to be prime for the two of them to be coprime?


No. 8 and 9 have a GCD of 1, so they are coprime even though neither of them is prime.

### Can we compute that probability?

Yes, and it won't take 24 hours to do it!

In [4]:
from math import gcd

for power in range(1, 2): # change 2 to 3 then 4.
  max = 10 ** power
  total = 0
  for m in range(1, max + 1):
    for n in range(1, max + 1):
      total += 1 if gcd(m, n) == 1 else 0

total

63

## The Euler Connection

In [ ]:
from math import pi
6 / (pi ** 2)

0.6079271018540267

### What? How??

From Donald Knuth's *The Art of Computer Programming* section 4.5.2:

The theorem, originally published and proven by Lejeune Dirichlet, says:

If $u$ and $v$ are integers chosen at random, the probability that $gcd(u, v) = 1$ is $$\frac{6}{\pi^2} \approx 0.60793.$$

## Hand Waving Proof

A precise formulation of this theorem, which carefully defines what is meant by being "chosen at random", [needs a rigorous proof]. Let us content ourselves here with a heuristic argument that shows why the theorem is plausible.

### Assumption

If we assume, without proof, the existence of a well-defined probability $p$ that $gcd(u, v)$ equals unity, then we can determine the probability that $gcd(u, v) = d$ for any positive integer $d$.

#### Why?

Because $gcd(u, v) = d$

if and only if

$u$ is a multiple of $d$ [i.e., $d$ is a divisor of $u$]

and

$v$ is a multiple of $d$ [i.e., $d$ is a divisor of $v$]

and

$gcd(u/d, v/d) = 1$.


### Which means?

 Thus the probability that $gcd(u, v) = d$ is equal to $1/d$ times $1/d$ times $p$, namely $p/d^2$.

### And so?

Now let us sum these probabilities over all possible values of $d$; we should get
$$1 = \sum_{d \ge 1} \frac{p}{d^2} = p\left(\frac{1}{1} + \frac{1}{4} + \frac{1}{9} + \frac{1}{16} + \cdots\right).$$


### Got it!!

Since the sum $$\left(\frac{1}{1} + \frac{1}{4} + \frac{1}{9} + \frac{1}{16} + \cdots\right)$$ is equal to $$\frac{\pi^2}{6},$$ we need $$p = \frac{6}{\pi^2}$$ in order to make this equation come out right.

# TODO Deduce Some Properties of Probabilities


## Now for your DPC task:
Remember that probability distributions satisfy two conditions:


1. $0 \le P(s) \le 1$ for each $s \in S$, and
2. $\sum_{s \in S} P(s) = 1$.

Knowing these two conditions and how the $s$ events partition the probability space $S$, we can infer these three additional facts:

1. For the **complementary** event, $P(\overline{s}) = 1 - P(s)$.
2. For the **union** of two events (either one of both happen), $P(E \cup F) = P(E) + P(F) - P(E \cap F)$.
3. For **independent events**, $P(E \cap F) = P(E)P(F)$.

Furthermore, the probability of the union of two independent events is the sum of the probabilities of the two events minus the product of their probabilities.

Given the above, what can you say about the probability of the union of two **mutually exclusive** events? What is the formula for that?





In [4]:
 ###P(A|B)??
  P(A and B) = 0
  P(A or B) = P(A) + P(B)
# 